lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -975,6 +982,79 @@ public void onAuthUpdated(final String token, final boolean waitForResponse) thr
+                    /* (RTC8a) If the connection is in the CONNECTED currentState and
+                     * auth.authorize is called or Ably requests a re-authentication
+                     * (see RTN22), the client must obtain a new token, then send an
+                     * AUTH ProtocolMessage to Ably with an auth attribute
+                     * containing an AuthDetails object with the token string. */
+                    try {
+                        ProtocolMessage msg = new ProtocolMessage(ProtocolMessage.Action.auth);
+                        msg.auth = new ProtocolMessage.AuthDetails(token);
+                        send(msg, false, null);
+                    } catch (AblyException e) {
+                        /* The send failed. Close the transport; if a subsequent
+                         * reconnect succeeds, it will be with the new token. */
+                        Log.v(TAG, "onAuthUpdated: closing transport after send failure");
+                        transport.close();
+                    }
+                    break;
+
+                case connecting:
+                    /* Close the connecting transport. */
+                    Log.v(TAG, "onAuthUpdated: closing connecting transport");
+                    ErrorInfo disconnectError = new ErrorInfo("Aborting incomplete connection with superseded auth params", 503, 80003);
+                    requestState(new StateIndication(ConnectionState.disconnected, disconnectError, null, null));
+                    /* Start a new connection attempt. */
+                    connect();
+                    break;
+
+                default:
+                    /* Start a new connection attempt. */
+                    connect();
+                    break;
+            }
+
+            /* Wait for a currentState transition into anything other than connecting or
+             * disconnected in a background thread */
+            singleThreadExecutor.execute(() -> {
+                boolean waitingForConnected = true;
+                while (waitingForConnected) {
+                    final ErrorInfo reason = waiter.waitForChange();
+                    final ConnectionState connectionState = currentState.state;
+                    switch (connectionState) {
+                        case connected:
+                            Log.v(TAG, "onAuthUpdated: got connected");
+                            waitingForConnected = false;
+                            break;
+
+                        case connecting:
+                        case disconnected:
+                            Log.v(TAG, "onAuthUpdated: " + connectionState);
+                            break;
+
+                        default:
+                            /* suspended/closed/error: throw the error. */
+                            Log.v(TAG, "onAuthUpdated: throwing exception");
+                            authUpdateResult.onUpdate(false, reason);
+                            waitingForConnected = false;
```

Comment:
I know that I'm the one who advised to put this instruction here but as I look at it now I see that's not enough :sweat_smile: This will break the `while` loop and jump to the instruction after the loop which is `authUpdateResult.onUpdate(true, null);`, so the `onUpdate()` callback will be called twice, once with an error and once with a success. We have 2 options here:
1. Put a `return` statement here and exit the function (but I'm not sure if the `finally` block would be called in that case)
2. Move the `authUpdateResult.onUpdate(true, null);` from line 1051 to the `case connected:` as we want to notify about the success once we are connected

I'd personally go with 2 :wink:

Reply:
Great spot. I made a change here : 1bbbd8ba1617fedc5fbd0cb0afb07e68fa676c04
Second one sounds more reasonable to me. But the first one would also work, because `finally` is executed regardless of you return from a function or not